src/crypto/secretstream/secretstream_macros.rs

```diff
@@ -196,6 +196,33 @@ impl Stream<Push> {
     /// message `m` and its `tag`. Optionally includes additional data `ad`,
     /// which is not encrypted.
     pub fn push(&mut self, m: &[u8], ad: Option<&[u8]>, tag: Tag) -> Result<Vec<u8>, ()> {
+        let buf_len = self.push_check(m, tag)?;
+
+        let mut buf = Vec::with_capacity(buf_len);
+        self.push_impl(m, ad, tag, &mut buf)?;
+        Ok(buf)
+    }
+
+    /// All data (including optional fields) is authenticated. Encrypts a
+    /// message `m` and its `tag`. Optionally includes additional data `ad`,
+    /// which is not encrypted.
+    ///
+    /// The encrypted message is written to the `out` vector, overwriting any existing data there.
+    pub fn push_to_vec(&mut self, m: &[u8], ad: Option<&[u8]>, tag: Tag, out: &mut Vec<u8>) -> Result<(), ()> {
+        let buf_len = self.push_check(m, tag)?;
+        out.clear();
+        out.reserve(buf_len);
+        self.push_impl(m, ad, tag, out)
+    }
+
+    /// Create a ciphertext for an empty message with the `TAG_FINAL` added
+    /// to signal the end of the stream. Since the `Stream` is not usable
+    /// after this point, this method consumes the `Stream.
+    pub fn finalize(mut self, ad: Option<&[u8]>) -> Result<Vec<u8>, ()> {
```

Comment:
Why did you move this method from the bottom of the impl into the middle?

Reply:
Because I put the private methods at the end. However, I'd be happy to move it back down to the bottom if you think that is better style.

Reply:
I see. We don't have any style for putting private methods at the end.